src/BinSkim.Sdk/CompilerDataLogger.cs

```diff
@@ -0,0 +1,172 @@
+
+        public CompilerDataLogger(IAnalysisContext analysisContext, string repositoryUri, string pipelineName)
+        {
+            try
+            {
+                string appInsightsKey = Environment.GetEnvironmentVariable("BinskimAppInsightsKey");
+                if (!string.IsNullOrEmpty(appInsightsKey) && Guid.TryParse(appInsightsKey, out _))
+                {
+                    Initialize(appInsightsKey);
+                    appInsightsRegistered = true;
+                }
+            }
+            catch (SecurityException)
+            {
+                // User does not have access to retrieve information from environment variables.
+            }
+
+            this.pipelineName = pipelineName;
+            this.repositoryUri = repositoryUri;
+            this.sha256 = analysisContext?.Hashes?.Sha256;
+            this.localPath = analysisContext?.TargetUri?.LocalPath;
+        }
+
+        public static void Initialize(string instrumentationKey)
+        {
+            s_telemetryConfiguration = new TelemetryConfiguration(instrumentationKey);
+            s_telemetryClient = new TelemetryClient(s_telemetryConfiguration);
+        }
+
+        public static void Flush()
+        {
+            s_telemetryClient?.Flush();
+        }
+
+        public void PrintHeader()
+        {
+            if (!appInsightsRegistered)
+            {
+                Console.WriteLine("RepositoryUri,PipelineName,Target,Compiler Name,Compiler BackEnd Version,Compiler FrontEnd Version,Language,Dialect,Module Name,Module Library,Hash,Error");
+            }
+        }
+
+        public void Write(string compilerData, ObjectModuleDetails omDetails)
+        {
+            string name = omDetails?.Name?.Replace(",", "_");
+            string library = omDetails?.Library?.Replace(",", ";");
+            if (appInsightsRegistered)
+            {
+                string[] compilerDataParts = compilerData.Split(',');
+
+                s_telemetryClient.TrackEvent(CompilerEventName, properties: new Dictionary<string, string>
+                {
+                    { "repositoryUri", repositoryUri ?? string.Empty },
+                    { "pipelineName", pipelineName ?? string.Empty },
+                    { "target", this.localPath },
+                    { "compilerName", compilerDataParts[0] },
+                    { "compilerBackEndVersion", compilerDataParts[1] },
+                    { "compilerFrontEndVersion", compilerDataParts[2] },
+                    { "language", compilerDataParts[3] },
+                    { "dialect", string.Empty },
+                    { "moduleName", name ?? string.Empty },
+                    { "moduleLibrary", (name == library ? string.Empty : library) },
+                    { "hash", this.sha256 ?? string.Empty },
+                    { "error", string.Empty }
+                });
+            }
+            else
+            {
+                Console.Write($"{repositoryUri},");
```

Comment:
>Write [](http://example.com/codeflow?start=24&length=5)

Note that it'd be more performant to concatenate all this information into a single string and then pass it to WriteLine.

Reply:
fixed